<a href="https://colab.research.google.com/github/purnaa-l/MiniProject-ML-Models/blob/main/Temperature%20versus%20AQI/DT_Regression_(AQI_versus_Temperature).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Decision Tree Regression

## Importing the libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the dataset

In [5]:
df = pd.read_csv('IndianWeatherRepository.csv')
df.head()
# X = dataset.iloc[:, :-1].values
# y = dataset.iloc[:, -1].values

country location_name          region  latitude  longitude      timezone  \
0   India    Ashoknagar  Madhya Pradesh     24.57      77.72  Asia/Kolkata   
1   India        Raisen  Madhya Pradesh     23.33      77.80  Asia/Kolkata   
2   India    Chhindwara  Madhya Pradesh     22.07      78.93  Asia/Kolkata   
3   India         Betul  Madhya Pradesh     21.86      77.93  Asia/Kolkata   
4   India   Hoshangabad  Madhya Pradesh     22.75      77.72  Asia/Kolkata   

   last_updated_epoch      last_updated  temperature_celsius  \
0          1693286100  2023-08-29 10:45                 27.5   
1          1693286100  2023-08-29 10:45                 27.5   
2          1693286100  2023-08-29 10:45                 26.3   
3          1693286100  2023-08-29 10:45                 25.6   
4          1693286100  2023-08-29 10:45                 27.2   

   temperature_fahrenheit  ... air_quality_PM2.5  air_quality_PM10  \
0                    81.5  ...              12.6              18.5   
1                    81.5  ...              10.7              14.2   
2                    79.3  ...              16.8              20.7   
3                    78.1  ...               4.9               6.6   
4                    81.0  ...              11.4              14.8   

   air_quality_us-epa-index  air_quality_gb-defra-index   sunrise    sunset  \
0                         1                           2  05:59 AM  06:41 PM   
1                         1                           1  06:00 AM  06:40 PM   
2                         2                           2  05:56 AM  06:34 PM   
3                         1                           1  06:00 AM  06:38 PM   
4                         1                           1  06:01 AM  06:39 PM   

   moonrise   moonset      moon_phase  moon_illumination  
0  05:42 PM  03:38 AM  Waxing Gibbous                 93  
1  05:39 PM  03:41 AM  Waxing Gibbous                 93  
2  05:32 PM  03:39 AM  Waxing Gibbous                 93  
3  05:36 PM  03:43 AM  Waxing Gibbous                 93  
4  05:38 PM  03:42 AM  Waxing Gibbous                 93  

[5 rows x 42 columns]

In [6]:
df = df[['temperature_celsius', 'air_quality_PM2.5', 'air_quality_PM10',
         'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'region', 'last_updated']]


In [7]:
# Check missing values
print(df.isnull().sum())

# Drop rows with missing AQI or temperature
df.dropna(subset=['temperature_celsius', 'air_quality_PM2.5'], inplace=True)


temperature_celsius           0
air_quality_PM2.5             0
air_quality_PM10              0
air_quality_us-epa-index      0
air_quality_gb-defra-index    0
region                        0
last_updated                  0
dtype: int64


In [8]:
df['hour'] = pd.to_datetime(df['last_updated']).dt.hour
df['day'] = pd.to_datetime(df['last_updated']).dt.dayofweek
df['region'] = df['region'].astype('category').cat.codes  # Convert categorical to numerical


In [9]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X = scaler.fit_transform(df[['temperature_celsius']])
y = df['air_quality_PM2.5']  # or any AQI index


In [10]:
df.head()

temperature_celsius  air_quality_PM2.5  air_quality_PM10  \
0                 27.5               12.6              18.5   
1                 27.5               10.7              14.2   
2                 26.3               16.8              20.7   
3                 25.6                4.9               6.6   
4                 27.2               11.4              14.8   

   air_quality_us-epa-index  air_quality_gb-defra-index  region  \
0                         1                           2      19   
1                         1                           1      19   
2                         2                           2      19   
3                         1                           1      19   
4                         1                           1      19   

       last_updated  hour  day  
0  2023-08-29 10:45    10    1  
1  2023-08-29 10:45    10    1  
2  2023-08-29 10:45    10    1  
3  2023-08-29 10:45    10    1  
4  2023-08-29 10:45    10    1

## Splitting the dataset into the Training set and Test set

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Training the Decision Tree Regression model on the Training set

In [12]:
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train, y_train)

DecisionTreeRegressor(random_state=0)

## Predicting the Test set results

In [14]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
# print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

## Evaluating the Model Performance

In [15]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.08311891404091698

The accuracy for Decision Tree is barely 0.083 which is not a reliable indicator at all.